Update Readme (#116)
virtee/snphost: README.md

# snphost

`snphost` is a Command Line Interface (CLI) utility designed for administrators managing AMD SEV-SNP enabled host systems. This tool facilitates interaction with the AMD SEV-SNP firmware device, enabling various operations such as certificate management and attestation processes.

- [Usage](#usage)
  - [1. help](#1-help)
  - [2. export](#2-export)
  - [3. import](#3-import)
  - [4. fetch ca](#4-fetch-ca)
  - [5. fetch vcek](#5-fetch-vcek)
  - [6. verify](#6-verify)
- [Building](#building)
  - [Ubuntu Dependencies](#ubuntu-dependencies)
  - [RHEL and Compatible Distributions Dependencies](#rhel-and-compatible-distributions-dependencies)
  - [openSUSE and Compatible Distributions Dependencies](#opensuse-and-compatible-distributions-dependencies)
- [Reporting Bugs](#reporting-bugs)

## Usage

### 1. help

Every `snphost` command and subcommand comes with a `--help` option that provides a description of its usage.

**Usage:**

```bash
snphost --help
```

or for a specific subcommand:

```bash
snphost <subcommand> --help
```

### 2. export

> Exports the SEV-SNP certificate chain from the AMD Platform Security Processor (PSP) to a specified directory. The user must specify the desired encoding format (`der` or `pem`), the certificate file name, and the target directory path. This command is useful for retrieving the certificate chain for attestation purposes.

Reviewer: ```suggestion
Deserializes a GHCB formatted cert chain file into individual certificates. The user must specify the desired encoding format (`der` or `pem`) for the certificates,  the file where the GHCB cert chain is stored in, and the target directory path where to store the deserialized certs. This command is useful for looking at the individual certs that will be used for an extended attestation.
```

The functionality changed in the latest iteration of Extended Attestation, the certs are no longer stored in the SP. We will probably see this functionality change again in the future.

**Usage:**

```bash
snphost export [der|pem] CERT-FILE DIR-PATH
```

**Example:**

```bash

> snphost export pem certificate.pem ./certs

Reviewer: ```suggestion
snphost export pem ghcb-certs.bin ./certs
```

```

### 3. import

> Imports serialized SEV-SNP certificates to a specified certificate file. This certificate file can then be provided to QEMU to perform extended attestation on guests. Currently, only the ASK, ARK, VCEK, and VLEK certificates are supported for serialization.

Reviewer: ```suggestion
Converts a certificate chain into a GHCB formatted file for extended attestation. This formatted file can then be provided to QEMU to perform extended attestation on guests. Currently, only the ASK, ASVK, ARK, VCEK, and VLEK certificates are supported for serialization.
```

**Usage:**

```bash
snphost import DIR-PATH CERT-FILE
```

**Example:**

```bash

> snphost import ./certs certificate.pem

Reviewer: ```suggestion
snphost import ./certs ghcb-certs.bin
```

```

### 4. fetch 

Command to request certificates from the KDS.

**Usage:**

```bash
snphost fetch <SUBCOMMAND>
```

**Subcommands**

#### 1. ca 

Fetches the host system's Certificate Authority (CA) certificate chain and writes the encoded certificates to the specified directory. Users must specify the desired encoding format (`der` or `pem`).

**Usage:**

```bash
snphost fetch ca [der|pem] DIR-PATH
```

**Example:**

```bash
snphost fetch ca pem ./certs
```

#### 2. vcek

Fetches the host system's Versioned Chip Endorsement Key (VCEK) and writes the encoded certificate to the specified directory. Users must specify the desired encoding format (`der` or `pem`).

**Usage:**

```bash
snphost fetch vcek [der|pem] DIR-PATH
```

**Example:**

```bash
snphost fetch vcek pem ./certs
```

### 6. verify

> Fetches certificates from the AMD PSP and verifies the certificate chain, ensuring its integrity and authenticity. This command is essential for validating the trustworthiness of the SEV-SNP environment.
> 

Reviewer: ```suggestion
Reads the certificates in a directory and verifies the certificate chain, ensuring its integrity and authenticity. This command is essential for validating the trustworthiness of the certificates that can be then passed to complete attestation.

```

**Usage:**

```bash

> snphost verify

Reviewer: ```suggestion
snphost verify ./certs
```

```

**Example:**

```bash
snphost fetch ca pem ./certs
```

#### 2. `vcek`

Fetches the host system's Versioned Chip Endorsement Key (VCEK) and writes the encoded certificate to the specified directory. Users must specify the desired encoding format (`der` or `pem`).

**Usage:**

```bash
snphost fetch vcek [der|pem] DIR-PATH
```

**Example:**

```bash
snphost fetch vcek pem ./certs
```

#### 3. `crl`
Fetches the latest Certificate Revocation List (CRL) from the configured source.

```sh
snphost fetch crl
```

**Example:**

```bash
snphost fetch crl
```

### 6. `show`
Displays various SNP-related information.

**Usage:**
```sh
snphost show <subcommands>
```

**Subcommands:**

#### 1. `guests`
Lists all active guests.

**Usage:**
```sh
snphost show guests
```

#### 2. `identifier`
Displays the SNP identifier.

**Usage:**
```sh
snphost show identifier
```

#### 3. `tcb`
Shows the current Trusted Computing Base (TCB) version.

**Usage:**
```sh
snphost show tcb
```

#### 4. `vcek-url`
Displays the URL for fetching the VCEK certificate.

**Usage:**
```sh
snphost show vcek-url
```

#### 5. `version`

Prints the current version of `snphost`.

**Usage:**
```sh
snphost show version

> ```

Reviewer: Were you able to to reflect on this, or do you think this is the best way to communicate this information?

### 7. `commit`

> Commits pending configuration changes.

Reviewer: ```suggestion
This command commits the current firmware and SNP platform config versions to the PSP.
This can't be undone and will not allow rollbacks to older versions.
```

Very Important that you clear up that this will not allow rollbacks (meaning you won't be able to change the config to a version prior to what has been committed to the ASP.

**Usage:**
```sh
snphost commit
```

### 8. `config`
Manages `snphost` configuration.

**Usage:**
```sh
snphost config <subcommand>
```

**Subcommands**

#### 1. `set`

> Sets a configuration parameter.

Reviewer: Still needs fixing

**Usage:**
```sh
snphost config set BOOTLOADER TEE SNP-FW MICROCODE MASK-CHIP
```

**Example:**
```sh

> snphost config set

Reviewer: Still needs fixing

```

#### 2.`reset`

> Resets a configuration parameter to its default value.

Reviewer: ```suggestion
This command resets the SEV-SNP platform. This will clear all
persistent data managed by the platform and reset the platform configuration
to its last committed version.
```

**Usage:**
```sh
snphost config reset 
```

> **Example:**
> ```sh
> snphost config reset
> ```

Reviewer: ```suggestion
```

**Usage:**

```bash
snphost import DIR-PATH CERT-FILE
```

**Example:**

```bash
snphost import ./certs ghcb-certs.bin
```
### 4. `ok`

Verifies if the `snpost` service is operational.

**Usage:**

```sh
snphost ok
```

### 5. `fetch`

Command to request certificates from the KDS.

**Usage:**

```bash
snphost fetch <subcommand>
```

**Subcommands:**

#### 1. `ca` 

Fetches the host system's Certificate Authority (CA) certificate chain and writes the encoded certificates to the specified directory. Users must specify the desired encoding format (`der` or `pem`).

**Usage:**

```bash
snphost fetch ca [der|pem] DIR-PATH
```

**Example:**

```bash
snphost fetch ca pem ./certs
```

#### 2. `vcek`

Fetches the host system's Versioned Chip Endorsement Key (VCEK) and writes the encoded certificate to the specified directory. Users must specify the desired encoding format (`der` or `pem`).

**Usage:**

```bash
snphost fetch vcek [der|pem] DIR-PATH
```

**Example:**

```bash
snphost fetch vcek pem ./certs
```

#### 3. `crl`
Fetches the latest Certificate Revocation List (CRL) from the configured source.

```sh
snphost fetch crl
```

**Example:**

```bash
snphost fetch crl
```

### 6. `show`

> Displays various SNP-related information.

Reviewer: Still needs to be resolved

**Usage:**
```sh
snphost show <subcommands>
```

**Subcommands:**

#### 1. `guests`
Lists all active guests.

**Usage:**
```sh
snphost show guests
```

#### 2. `identifier`

> Displays the SNP identifier.

Reviewer: Still needs resolving

**Usage:**
```sh
snphost show identifier
```

#### 3. `tcb`

> Shows the current Trusted Computing Base (TCB) version.

Reviewer: Still needs resolving

**Usage:**
```sh
snphost show tcb
```

#### 4. `vcek-url`
Displays the URL for fetching the VCEK certificate.

**Usage:**
```sh
snphost show vcek-url
```

#### 5. `version`

> Prints the current version of `snphost`.

Reviewer: Still needs resolving

```bash
git clone https://github.com/virtee/snphost.git
cargo build --release
cd snphost/target/release
```

The compiled binary will be located in the `target/release` directory.

## Common Workflows

This section outlines common workflows for configuring and using `snphost` effectively on a host system with AMD SEV-SNP enabled.

### 1. Changing the SNP Configuration

To change the configuration of the SNP firmware, use the `config set` command followed by `commit`.

```bash
snphost config set BOOTLOADER TEE SNP-FW MICROCODE MASK-CHIP
snphost commit
````

**What this does:**

* `config set`: Updates the in-memory configuration parameters.
* `commit`: Writes the updated configuration to persistent storage, making it active for the next boot or system event.

**Example:**

```bash
snphost config set grub sev 0x10000000 0x20000000 0xff
snphost commit
```

You can change settings such as:

* `BOOTLOADER`: e.g., `grub`, `systemd-boot`
* `TEE`: Trusted Execution Environment type, e.g., `sev`
* `SNP-FW`: Address of the SNP firmware
* `MICROCODE`: Microcode binary address
* `MASK-CHIP`: Bitmask controlling chip-specific behavior

> ---

Reviewer: None of these values accurately reflect what should be passed into the command. Config changes the reported versions of these parameters. I would  look more closely at the already existing documentation. Please look at both the abi and the snohost.1.adoc.

### 2. Resetting Config Changes

To discard uncommitted configuration changes and revert to the last committed state:

```bash
snphost config reset
```

**Workflow:**

```bash
snphost config set grub sev 0x10000000 0x20000000 0xff
# decide to discard changes
snphost config reset
```

This is useful for reverting accidental or experimental changes before they are made permanent.

---

### 3. Extended Attestation Flow

> This flow demonstrates fetching host certificates, exporting them for inspection, and passing them to QEMU for extended guest attestation.

Reviewer: ```suggestion
This flow demonstrates fetching host certificates, verifying the certs, importing them into the correct gchb format, and passing them to QEMU for extended guest attestation.
```

> This flow demonstrates fetching host certificates, exporting them for inspection, and passing them to QEMU for extended guest attestation.

Reviewer: I would also add a notice on this workflow saying something like:

"This was the workflow used to set-up the host for extended attestation, this functionality is currently unavailable in upstream kernel, and it's subject to change"

```bash
# Fetch certificates from KDS
snphost fetch ca pem ./certs
snphost fetch vcek pem ./certs

# Verify the certificate chain (optional but recommended)
snphost verify ./certs

# Serialize to GHCB format for QEMU
snphost import ./certs ghcb-certs.bin

# (Optional) To inspect GHCB-formatted file contents:
snphost export pem ghcb-certs.bin ./decoded-certs
```

**Use with QEMU:**

You can now pass `ghcb-certs.bin` to QEMU via the `sev-snp-certs` option:

```bash
qemu-system-x86_64 \
  ... \
  -object sev-snp-guest,id=sev0,sev-snp-certs=ghcb-certs.bin \
  ...
```

> This enables guest attestation using the extended certificate chain.

Reviewer: ```suggestion
The certificates can now be retrieved in the guest using extended attestation.
```

```

**Example Output:**

```bash
Reported TCB: TCB Version:
  Microcode:   25
  SNP:         23
  TEE:         0
  Boot Loader: 10
  FMC:         None

Platform TCB: TCB Version:
  Microcode:   25
  SNP:         23
  TEE:         0
  Boot Loader: 10
  FMC:         None
```
With the config command, you can change these values to an older version, but not to a newer version. Attempting to set a newer version will result in a failure.

#### Part 2. Mask Chip ID
The Mask Chip ID can be set to a value between 0 and 3, corresponding to its binary counterpart:
```bash
0 - 00
1 - 01
2 - 10
3 - 11
```
These values toggle the MASK-CHIP-KEY and MASK-CHIP-ID settings:

```bash
Bit 0: MASK-CHIP-ID
Bit 1: MASK-CHIP-KEY
```
For example, if you pass 2:

```bash
Bit 0 is 0, meaning MASK-CHIP-ID is disabled.
Bit 1 is 1, meaning MASK-CHIP-KEY is enabled.
```

To set the TCB parameters and Mask Chip ID, you can use the following command:

**Example:**
```bash
snphost config set 10 0 23 25 0
```

In this example,
```bash
10 is the Bootloader version.
0 is the TEE version.
23 is the SNP Firmware level.
25 is the Microcode level.
0 is the Mask Chip ID value.
```
This command will update the values.

> To commit these changes, you must use the snphost commit command by running:
> 
> ```bash
> snphost commit
> ```
> *Note: Commit changes config permanently. Do Not commit values unless you are absolutely sure.*

Reviewer: ```suggestion
To commit these changes to firmware, you must use the snphost commit command by running:

```bash
snphost commit
```
*Note: Commit changes config permanently. You can't roll back to your previous version. Do Not commit values unless you are absolutely sure.*
```

---

### 2. Resetting Config Changes

To discard uncommitted configuration changes and revert to the last committed state:

```bash
snphost config reset
```

**Workflow:**

```bash

> snphost config set grub sev 0x10000000 0x20000000 0xff

Reviewer: ???
